components/butterfly/src/rumor/dat_file.rs

```diff
@@ -56,138 +56,58 @@ const HEADER_VERSION_2_SIZE: usize =
 /// * Header Body - Variable bytes - see Header
 /// * Rumors - Variable bytes
 #[derive(Debug)]
-pub struct DatFile {
-    header:      Header,
-    header_size: u64,
-    path:        PathBuf,
-    reader:      BufReader<File>,
+struct DatFile {
+    header: Header,
+    path:   PathBuf,
 }
 
-impl DatFile {
+#[derive(Debug)]
+pub struct DatFileReader {
+    dat_file: DatFile,
+    reader:   BufReader<File>,
+}
+
+#[derive(Debug)]
+pub struct DatFileWriter(DatFile);
+
+impl DatFileReader {
     /// # Locking
     /// * `MemberList::entries` (read) This method must not be called while any MemberList::entries
     ///   lock is held.
     pub fn read_or_create_mlr(data_path: PathBuf, server: &Server) -> Result<Self> {
-        let file = OpenOptions::new().create(true)
+        let size = OpenOptions::new().create(true)
                                      .read(true)
                                      .write(true)
                                      .open(&data_path)
-                                     .map_err(|err| Error::DatFileIO(data_path.clone(), err))?;
-        let size = file.metadata()
-                       .map_err(|err| Error::DatFileIO(data_path.clone(), err))?
-                       .len();
-        let reader = BufReader::new(file);
-        let mut dat_file = DatFile { path: data_path,
-                                     header_size: 0,
-                                     header: Header::default(),
-                                     reader };
+                                     .map_err(|err| Error::DatFileIO(data_path.clone(), err))?
+                                     .metadata()
+                                     .map_err(|err| Error::DatFileIO(data_path.clone(), err))?
+                                     .len();
+
+        let dat_file = DatFile { path:   data_path.clone(),
+                                 header: Header::default(), };
 
         if size == 0 {
-            dat_file.write_mlr(server)?;
-        } else {
-            dat_file.read_header()?;
+            DatFileWriter::new(data_path).write_mlr(server)?;
         }
 
-        Ok(dat_file)
+        Self::reader_creation(dat_file)
```

Comment:
I think this can be simplified to:
```rust
        if size == 0 {
            DatFileWriter::new(data_path.clone()).write_mlr(server)?;
        }

        Self::reader_creation(&data_path)
```
There's no need to create a `DatFile` at all.

I think I was a bit off in the original idea about what belongs in `DateFileReader` vs `DatFileWriter`. Poking at it a bit, I came to (`derive`s omitted for clarity):
```rust
struct DatFile(PathBuf);

pub struct DatFileReader {
    header:   Header,
    dat_file: DatFile,
    reader:   BufReader<File>,
}

pub struct DatFileWriter(DatFile);
```
The write case doesn't need to keep the header around. it's re-created in `write_mlr`, so having it as a field is rather misleading. In fact, it would be possible to eliminate `DatFile` altogether and just have a `PathBuf`, but I think it's nice to have somewhere to hang that documentation and show the conceptual linkage between `DateFileReader` vs `DatFileWriter`.

This does require shuffling things around a bit, but it wasn't too much effort to get things to build. [Here's a diff](https://github.com/habitat-sh/habitat/files/3352656/pr-6703.diff.zip), so we don't have to duplicate effort.

Reply:
Sounds good.  I'll make the appropriate changes.  Thanks for including a diff 😁